### PR TITLE
[C10D] call destroy_process_group after MultiProcess tests

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140820

Faced with an annoying string of warnings like this when running tests,
Screenshot 2024-11-15 at 11 23 21 AM

My choices seem to be (1) call destroy_process_group() at the end of
each test fn, (2) do this in some wrapper, (3) do it in the base test
class.

Since tests in MultiProcessTestCase are responsible for calling
init_process_group themselves, they should also be responsible for
calling destroy (or at least method (3) would be asymmetric and may
result in double-destroy).

But it doesn't feel worth it to go add a destroy call manually to each
test, and try/except for a possible second destroy call seems like a
happy middle ground.

Note: tests that want to ensure that destroy runs cleanly can and should
still call destroy _inside_ the test, and this change does not affect
that.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/testing/_internal/common_distributed.py [hunk 1]
```diff
@@ -714,6 +714,14 @@ def run_test(self, test_name: str, parent_pipe) -> None:
             # Close pipe after done with test.
             parent_pipe.close()
 
+        try:
+            # Many test cases init a process group but do not destroy it.
+            # Some tests do destroy the pgs, and destroy can't be called twice.
+            # This avoids spewing warnings about improperly shutting down.
+            c10d.destroy_process_group()
```
[the remark this answers is not in the document]
Comment (Skylion007) [reply torch/testing/_internal/common_distributed.py:721]: Unfortunately, we don't block on it in anyway so joining on the process does not guarantee that the PG is group is destroyed. destroy_process_group really only STARTS the process group destruction.

[the remark this answers is not in the document]
Comment (wconstab) [reply torch/testing/_internal/common_distributed.py:721]: i'm saying that all N processes are ensured to be killed before we start the next group of N processes.  Its the same thing as running your torchrun program in a loop, but doing `kill -9` in-between loop iterations.  Isn't that enough?

Alternatively, what are you saying we should do- never call destroy_process_group?

Note: i believe your point about not destroying and initializing within the same process. But we are not doing that here.

**Review by fegin — APPROVED** [review]
dtensor test also calls destroy in the end of the test. So I think it is safe for multi-processing tests. Not sure if it worths to call a cuda synchronize if cuda is available. I had some issues that NCCL will crash if I don't call synchronize before calling destroying PG, but I could not always reproduce the issue.

**Pull request closed without merging**